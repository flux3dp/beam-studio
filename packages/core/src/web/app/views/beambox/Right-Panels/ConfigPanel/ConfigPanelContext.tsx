import { createContext } from 'react';

interface Context {
  selectedLayers: string[];
  simpleMode?: boolean;
}

export default createContext<Context>({} as Context);
