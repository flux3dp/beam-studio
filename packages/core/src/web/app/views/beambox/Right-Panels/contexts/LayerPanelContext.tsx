import React, { createContext, useCallback, useEffect, useMemo, useState } from 'react';

import { pipe } from 'remeda';

import { promarkModels } from '@core/app/actions/beambox/constant';
import { LayerModule } from '@core/app/constants/layer-module/layer-modules';
import eventEmitterFactory from '@core/helpers/eventEmitterFactory';
import useWorkarea from '@core/helpers/hooks/useWorkarea';
import doLayersContainsVector from '@core/helpers/layer/check-vector';
import { getData } from '@core/helpers/layer/layer-config-helper';
import { getLayerElementByName } from '@core/helpers/layer/layer-helper';
import useForceUpdate from '@core/helpers/use-force-update';

interface ILayerPanelContext {
  forceUpdate: () => void;
  forceUpdateSelectedLayers: () => void;
  hasGradient: boolean;
  hasVector: boolean;
  selectedLayers: string[];
  setSelectedLayers: (selectedLayers: string[]) => void;
}

export const LayerPanelContext = createContext<ILayerPanelContext>({
  forceUpdate: () => {},
  forceUpdateSelectedLayers: () => {},
  hasGradient: false,
  hasVector: false,
  selectedLayers: [],
  setSelectedLayers: () => {},
});

const layerPanelEventEmitter = eventEmitterFactory.createEventEmitter('layer-panel');

interface Props {
  children?: React.ReactNode;
}

export const LayerPanelContextProvider = ({ children }: Props): React.JSX.Element => {
  const [hasVector, setHasVector] = useState<boolean>(false);
  const [hasGradient, setHasGradient] = useState<boolean>(false);
  const [selectedLayers, setSelectedLayers] = useState<string[]>([]);
  const forceUpdate = useForceUpdate();
  const forceUpdateSelectedLayers = useCallback(() => setSelectedLayers([...selectedLayers]), [selectedLayers]);
  const workarea = useWorkarea();
  const isPromark = useMemo(() => promarkModels.has(workarea), [workarea]);
  const lazySetSelectedLayers = useCallback(
    (newLayers: string[]) => {
      if (newLayers.length === selectedLayers.length && newLayers.every((name, i) => name === selectedLayers[i])) {
        return;
      }

      const isUvExportable = newLayers.every(
        (layerName) => getData(getLayerElementByName(layerName), 'module') === LayerModule.UV_EXPORT,
      );

      layerPanelEventEmitter.emit('updateUvExportStatus', isUvExportable);

      setSelectedLayers(newLayers);
    },
    [selectedLayers, setSelectedLayers],
  );

  useEffect(() => {
    layerPanelEventEmitter.on('UPDATE_LAYER_PANEL', forceUpdate);

    return () => {
      layerPanelEventEmitter.removeListener('UPDATE_LAYER_PANEL', forceUpdate);
    };
    // eslint-disable-next-line hooks/exhaustive-deps
  }, []);

  useEffect(() => {
    layerPanelEventEmitter.on('SET_SELECTED_LAYERS', lazySetSelectedLayers);

    return () => {
      layerPanelEventEmitter.removeListener('SET_SELECTED_LAYERS', lazySetSelectedLayers);
    };
  }, [lazySetSelectedLayers]);

  useEffect(() => {
    const getSelectedLayers = (response: { selectedLayers: string[] }) => {
      response.selectedLayers = selectedLayers;
    };

    layerPanelEventEmitter.on('GET_SELECTED_LAYERS', getSelectedLayers);

    const checkVector = () => {
      const newVal = doLayersContainsVector(selectedLayers);

      setHasVector(newVal);
    };

    layerPanelEventEmitter.on('CHECK_VECTOR', checkVector);

    return () => {
      layerPanelEventEmitter.off('GET_SELECTED_LAYERS', getSelectedLayers);
      layerPanelEventEmitter.off('CHECK_VECTOR', checkVector);
    };
  }, [selectedLayers]);

  useEffect(() => {
    const checkGradient = () => {
      if (isPromark) {
        const newVal = selectedLayers.some((layerName: string) =>
          pipe(
            //
            layerName,
            getLayerElementByName,
            (layer) => layer?.querySelector('image[data-shading="true"]'),
            Boolean,
          ),
        );

        setHasGradient(newVal);
      }
    };

    checkGradient();
    layerPanelEventEmitter.on('CHECK_GRADIENT', checkGradient);

    return () => {
      layerPanelEventEmitter.removeListener('CHECK_GRADIENT', checkGradient);
    };
  }, [selectedLayers, isPromark]);

  // move doLayersContainsVector and setVector in effect to avoid heavy render process
  // eslint-disable-next-line hooks/exhaustive-deps
  useEffect(() => {
    const newVal = doLayersContainsVector(selectedLayers);

    if (hasVector !== newVal) {
      setHasVector(newVal);
    }

    if (isPromark) {
      const newGradientVal = selectedLayers.some((layerName: string) =>
        pipe(
          //
          layerName,
          getLayerElementByName,
          (layer) => layer?.querySelector('image[data-shading="true"]'),
          Boolean,
        ),
      );

      if (newGradientVal !== hasGradient) {
        setHasGradient(newGradientVal);
      }
    }
  });

  return (
    <LayerPanelContext.Provider
      value={{
        forceUpdate,
        forceUpdateSelectedLayers,
        hasGradient,
        hasVector,
        selectedLayers,
        setSelectedLayers: lazySetSelectedLayers,
      }}
    >
      {children}
    </LayerPanelContext.Provider>
  );
};
