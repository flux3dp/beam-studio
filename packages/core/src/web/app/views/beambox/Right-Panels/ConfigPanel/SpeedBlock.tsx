import React, { memo, useContext, useMemo } from 'react';

import { ExclamationCircleOutlined, QuestionCircleOutlined } from '@ant-design/icons';
import { Tooltip } from 'antd';
import { Button, Popover } from 'antd-mobile';
import classNames from 'classnames';
import { sprintf } from 'sprintf-js';
import { useShallow } from 'zustand/react/shallow';

import { promarkModels } from '@core/app/actions/beambox/constant';
import { getAddOnInfo } from '@core/app/constants/addOn';
import { getSpeedOptions } from '@core/app/constants/config-options';
import { printingModules } from '@core/app/constants/layer-module/layer-modules';
import { getWorkarea } from '@core/app/constants/workarea-constants';
import { useConfigPanelStore } from '@core/app/stores/configPanel';
import { useDocumentStore } from '@core/app/stores/documentStore';
import { useGlobalPreferenceStore } from '@core/app/stores/globalPreferenceStore';
import history from '@core/app/svgedit/history/history';
import undoManager from '@core/app/svgedit/history/undoManager';
import { LayerPanelContext } from '@core/app/views/beambox/Right-Panels/contexts/LayerPanelContext';
import { ObjectPanelContext } from '@core/app/views/beambox/Right-Panels/contexts/ObjectPanelContext';
import ObjectPanelItem from '@core/app/views/beambox/Right-Panels/ObjectPanelItem';
import objectPanelItemStyles from '@core/app/views/beambox/Right-Panels/ObjectPanelItem.module.scss';
import { getAutoFeeder } from '@core/helpers/addOn';
import eventEmitterFactory from '@core/helpers/eventEmitterFactory';
import useHasCurveEngraving from '@core/helpers/hooks/useHasCurveEngraving';
import useWorkarea from '@core/helpers/hooks/useWorkarea';
import doLayersContainsVector from '@core/helpers/layer/check-vector';
import { CUSTOM_PRESET_CONSTANT, writeData } from '@core/helpers/layer/layer-config-helper';
import round from '@core/helpers/math/round';
import units from '@core/helpers/units';
import useI18n from '@core/helpers/useI18n';
import storage from '@core/implementations/storage';

import styles from './Block.module.scss';
import ConfigPanelContext from './ConfigPanelContext';
import ConfigSlider from './ConfigSlider';
import ConfigValueDisplay from './ConfigValueDisplay';
import initState from './initState';

const SpeedBlock = ({ type = 'default' }: { type?: 'default' | 'modal' | 'panel-item' }): React.JSX.Element => {
  const lang = useI18n();
  const t = lang.beambox.right_panel.laser_panel;
  const { change, module, speed } = useConfigPanelStore();
  const { selectedLayers } = useContext(ConfigPanelContext);
  const simpleMode = !useGlobalPreferenceStore((state) => state['print-advanced-mode']);
  const { activeKey } = useContext(ObjectPanelContext);
  const visible = activeKey === 'speed';
  const { hasVector } = useContext(LayerPanelContext);
  const hasCurveEngraving = useHasCurveEngraving();
  const timeEstimationButtonEventEmitter = useMemo(
    () => eventEmitterFactory.createEventEmitter('time-estimation-button'),
    [],
  );

  const { hasMultiValue, value } = speed;
  const layerModule = module.value;
  const isPrinting = useMemo(() => printingModules.has(layerModule), [layerModule]);

  const {
    calculateUnit: fakeUnit,
    decimal,
    display: displayUnit,
  } = useMemo(() => {
    const unit: 'inches' | 'mm' = storage.get('default-units') || 'mm';
    const display = { inches: 'in/s', mm: 'mm/s' }[unit];

    const calculateUnit: 'inch' | 'mm' = { inches: 'inch', mm: 'mm' }[unit] as any;
    const d = { inches: 2, mm: 1 }[unit];

    return { calculateUnit, decimal: d, display };
  }, []);
  const workarea = useWorkarea();
  const isPromark = useMemo(() => promarkModels.has(workarea), [workarea]);
  const addOnInfo = useMemo(() => getAddOnInfo(workarea), [workarea]);
  const { autoFeeder, borderless } = useDocumentStore(
    useShallow((state) => ({
      autoFeeder: state['auto-feeder'],
      borderless: state.borderless,
    })),
  );
  const isAutoFeederOn = useMemo(
    () => getAutoFeeder(addOnInfo, { autoFeeder, borderless }),
    [addOnInfo, autoFeeder, borderless],
  );
  const {
    curveSpeedLimit,
    maxSpeed: maxValue,
    minSpeed: minValue,
    minSpeedWarning,
    vectorSpeedLimit,
  } = useMemo(() => {
    const workareaObj = getWorkarea(workarea);

    return {
      curveSpeedLimit: workareaObj.curveSpeedLimit,
      maxSpeed: workareaObj.maxSpeed,
      minSpeed: workareaObj.minSpeed,
      minSpeedWarning: workareaObj.minSpeedWarning,
      vectorSpeedLimit: (isAutoFeederOn && addOnInfo.autoFeeder?.vectorSpeedLimit) || workareaObj.vectorSpeedLimit,
    };
  }, [workarea, addOnInfo, isAutoFeederOn]);

  const curveEngravingSpeedWarning = useMemo(() => {
    if (!curveSpeedLimit) {
      return '';
    }

    return sprintf(t.speed_constrain_warning_curve_engraving, {
      limit: fakeUnit === 'mm' ? `${curveSpeedLimit} mm/s` : `${round(curveSpeedLimit / 25.4, 2)} in/s`,
    });
  }, [fakeUnit, curveSpeedLimit, t.speed_constrain_warning_curve_engraving]);

  const vectorSpeedWarning = useMemo(() => {
    if (!vectorSpeedLimit) {
      return '';
    }

    return sprintf(isAutoFeederOn ? t.speed_constrain_warning_auto_feeder : t.speed_constrain_warning, {
      limit: fakeUnit === 'mm' ? `${vectorSpeedLimit} mm/s` : `${round(vectorSpeedLimit / 25.4, 2)} in/s`,
    });
  }, [fakeUnit, t, vectorSpeedLimit, isAutoFeederOn]);
  const hasCurveLimit = useGlobalPreferenceStore((state) => state.curve_engraving_speed_limit);
  const hasVectorLimit = useGlobalPreferenceStore((state) => state.vector_speed_constraint);

  let warningText = '';

  if (!isPromark) {
    if (hasCurveLimit && hasCurveEngraving && curveSpeedLimit && value > curveSpeedLimit) {
      warningText = curveEngravingSpeedWarning;
    } else if (hasVector && hasVectorLimit && vectorSpeedLimit && value > vectorSpeedLimit) {
      warningText = vectorSpeedWarning;
    } else if (minSpeedWarning && value < minSpeedWarning) {
      warningText = t.low_speed_warning;
    }
  }

  const handleChange = (val: number) => {
    change({ configName: CUSTOM_PRESET_CONSTANT, speed: val });
    timeEstimationButtonEventEmitter.emit('SET_ESTIMATED_TIME', null);

    if (type !== 'modal') {
      const batchCmd = new history.BatchCommand('Change speed');

      selectedLayers.forEach((layerName) => {
        writeData(layerName, 'speed', val, { applyPrinting: true, batchCmd });
        writeData(layerName, 'configName', CUSTOM_PRESET_CONSTANT, { batchCmd });
      });
      batchCmd.onAfter = initState;
      undoManager.addCommandToHistory(batchCmd);
    }
  };

  const sliderOptions = useMemo(
    () => (simpleMode && isPrinting ? getSpeedOptions(lang, layerModule) : undefined),
    [simpleMode, isPrinting, lang, layerModule],
  );

  const content = (
    <div className={classNames(styles.panel, styles[type])}>
      <span className={styles.title}>
        {t.speed}
        {isPromark && (
          <Tooltip title={t.promark_speed_desc}>
            <QuestionCircleOutlined className={styles.hint} />
          </Tooltip>
        )}
      </span>
      <ConfigValueDisplay
        decimal={decimal}
        hasMultiValue={hasMultiValue}
        inputId="speed-input"
        max={maxValue}
        min={minValue}
        onChange={handleChange}
        options={sliderOptions}
        type={type}
        unit={displayUnit}
        value={value}
      />
      <ConfigSlider
        decimal={decimal}
        id="speed"
        max={maxValue}
        min={minValue}
        onChange={handleChange}
        options={sliderOptions}
        speedLimit={!isPrinting && (type === 'modal' ? doLayersContainsVector(selectedLayers) : hasVector)}
        step={0.1}
        unit={displayUnit}
        value={value}
      />
      {warningText ? (
        <div className={styles.warning}>
          <div className={styles['warning-icon']}>
            <ExclamationCircleOutlined />
          </div>
          <div className={styles['warning-text']}>{warningText}</div>
        </div>
      ) : null}
    </div>
  );

  const displayValue = useMemo(() => {
    const selectedOption = sliderOptions?.find((opt) => opt.value === value);

    if (selectedOption) {
      return selectedOption.label;
    }

    return +units.convertUnit(value, fakeUnit, 'mm')?.toFixed(decimal);
  }, [decimal, sliderOptions, value, fakeUnit]);

  return type === 'panel-item' ? (
    <Popover content={content} visible={visible}>
      <ObjectPanelItem.Item
        autoClose={false}
        content={
          <Button
            className={classNames(objectPanelItemStyles['number-item'], styles['display-btn'])}
            fill="outline"
            shape="rounded"
            size="mini"
          >
            <span style={{ whiteSpace: 'nowrap' }}>{displayValue}</span>
          </Button>
        }
        id="speed"
        label={t.speed}
      />
    </Popover>
  ) : (
    content
  );
};

export default memo(SpeedBlock);
