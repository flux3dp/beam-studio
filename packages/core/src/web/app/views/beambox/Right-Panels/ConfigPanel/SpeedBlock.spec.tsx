import React from 'react';

import { fireEvent, render } from '@testing-library/react';

import LayerModule from '@core/app/constants/layer-module/layer-modules';
import { LayerPanelContext } from '@core/app/views/beambox/Right-Panels/contexts/LayerPanelContext';

import ConfigPanelContext from './ConfigPanelContext';

jest.mock('./ConfigSlider', () => ({ id, max, min, onChange, value }: any) => (
  <input
    className="mock-config-slider"
    id={id}
    max={max}
    min={min}
    onChange={(e) => onChange(Number(e.target.value))}
    step={0.1}
    type="range"
    value={value}
  />
));

jest.mock(
  './ConfigValueDisplay',
  () =>
    ({
      decimal = 0,
      hasMultiValue = false,
      inputId,
      max,
      min,
      onChange,
      options,
      type = 'default',
      unit,
      value,
    }: any) => (
      <div>
        MockConfigValueDisplay
        <p>inputId: {inputId}</p>
        <p>type: {type}</p>
        <p>max: {max}</p>
        <p>min: {min}</p>
        <p>value: {value}</p>
        <p>unit: {unit}</p>
        <p>hasMultiValue: {hasMultiValue}</p>
        <p>decimal: {decimal}</p>
        <p>options: {JSON.stringify(options)}</p>
        <button onClick={() => onChange(88)} type="button">
          MockConfigValueDisplayButton
        </button>
      </div>
    ),
);

jest.mock('@core/app/views/beambox/Right-Panels/ObjectPanelItem', () => ({
  Item: ({ content, id, label, onClick }: any) => (
    <div onClick={onClick}>
      MockObjectPanelItem
      <div>id: {id}</div>
      <div>content: {content}</div>
      <div>label: {label}</div>
    </div>
  ),
}));

const mockWriteData = jest.fn();

jest.mock('@core/helpers/layer/layer-config-helper', () => ({
  CUSTOM_PRESET_CONSTANT: 'CUSTOM_PRESET_CONSTANT',
  writeData: (...args) => mockWriteData(...args),
}));

const mockStorageGet = jest.fn();

jest.mock('@core/implementations/storage', () => ({
  get: (...args) => mockStorageGet(...args),
}));

const mockPrefRead = jest.fn();

jest.mock('@core/app/actions/beambox/beambox-preference', () => ({
  read: (...args) => mockPrefRead(...args),
}));

const mockAddCommandToHistory = jest.fn();

jest.mock('@core/helpers/svg-editor-helper', () => ({
  getSVGAsync: (callback) =>
    callback({
      Canvas: {
        addCommandToHistory: mockAddCommandToHistory,
      },
    }),
}));

let batchCmd = { count: 0, onAfter: undefined };
const mockBatchCommand = jest.fn().mockImplementation(() => {
  batchCmd = { count: batchCmd.count + 1, onAfter: undefined };

  return batchCmd;
});

jest.mock('@core/app/svgedit/history/history', () => ({
  BatchCommand: mockBatchCommand,
}));

const mockSelectedLayers = ['layer1', 'layer2'];
const mockContextState = {
  module: { hasMultiValue: false, value: LayerModule.LASER_10W_DIODE },
  speed: { hasMultiValue: false, value: 87 },
};
const mockDispatch = jest.fn();
const mockInitState = jest.fn();

const mockCreateEventEmitter = jest.fn();

jest.mock('@core/helpers/eventEmitterFactory', () => ({
  createEventEmitter: (...args) => mockCreateEventEmitter(...args),
}));

const mockEmit = jest.fn();

jest.mock('@core/app/views/beambox/Right-Panels/contexts/LayerPanelContext', () => ({
  LayerPanelContext: React.createContext({ hasVector: false }),
}));

jest.mock('@core/app/views/beambox/Right-Panels/contexts/ObjectPanelContext', () => ({
  ObjectPanelContext: React.createContext({ activeKey: null }),
}));

jest.mock('@core/helpers/layer/check-vector', () => jest.fn());

import SpeedBlock from './SpeedBlock';

const mockUseHasCurveEngraving = jest.fn();

jest.mock('@core/helpers/hooks/useHasCurveEngraving', () => () => mockUseHasCurveEngraving());

const mockUseWorkarea = jest.fn();

jest.mock('@core/helpers/hooks/useWorkarea', () => () => mockUseWorkarea());

const mockUseAutoFeeder = jest.fn();

jest.mock('@core/helpers/addOn', () => ({
  useAutoFeeder: (...args) => mockUseAutoFeeder(...args),
}));

describe('test SpeedBlock', () => {
  beforeEach(() => {
    jest.clearAllMocks();
    mockCreateEventEmitter.mockReturnValueOnce({
      emit: mockEmit,
    });
    mockUseHasCurveEngraving.mockReturnValue(false);
    mockStorageGet.mockReturnValue('mm');
    mockPrefRead.mockReturnValueOnce(true).mockReturnValueOnce(true);
    mockUseWorkarea.mockReturnValue('fbm1');
    mockUseAutoFeeder.mockReturnValue(false);
  });

  it('should render correctly when unit is mm', () => {
    const { container } = render(
      <ConfigPanelContext.Provider
        value={{
          dispatch: mockDispatch,
          initState: mockInitState,
          selectedLayers: mockSelectedLayers,
          state: mockContextState as any,
        }}
      >
        <SpeedBlock />
      </ConfigPanelContext.Provider>,
    );

    expect(mockStorageGet).toHaveBeenCalledTimes(1);
    expect(mockStorageGet).toHaveBeenLastCalledWith('default-units');
    expect(mockPrefRead).toHaveBeenCalledTimes(2);
    expect(mockPrefRead).toHaveBeenNthCalledWith(1, 'curve_engraving_speed_limit');
    expect(mockPrefRead).toHaveBeenNthCalledWith(2, 'vector_speed_constraint');
    expect(container).toMatchSnapshot();
  });

  it('should render correctly when unit is inches', () => {
    mockStorageGet.mockReturnValueOnce('inches');

    const { container } = render(
      <ConfigPanelContext.Provider
        value={{
          dispatch: mockDispatch,
          initState: mockInitState,
          selectedLayers: mockSelectedLayers,
          state: mockContextState as any,
        }}
      >
        <SpeedBlock />
      </ConfigPanelContext.Provider>,
    );

    expect(container).toMatchSnapshot();
  });

  it('should render correctly when type is panel-item', () => {
    const { container } = render(
      <ConfigPanelContext.Provider
        value={{
          dispatch: mockDispatch,
          initState: mockInitState,
          selectedLayers: mockSelectedLayers,
          state: mockContextState as any,
        }}
      >
        <SpeedBlock type="panel-item" />
      </ConfigPanelContext.Provider>,
    );

    expect(container).toMatchSnapshot();
  });

  it('should render correctly when has vector warning', () => {
    const { container } = render(
      <LayerPanelContext.Provider value={{ hasVector: true } as any}>
        <ConfigPanelContext.Provider
          value={{
            dispatch: mockDispatch,
            initState: mockInitState,
            selectedLayers: mockSelectedLayers,
            state: mockContextState as any,
          }}
        >
          <SpeedBlock />
        </ConfigPanelContext.Provider>
      </LayerPanelContext.Provider>,
    );

    expect(container).toMatchSnapshot();
    expect(container.querySelector('.warning')).toBeInTheDocument();
  });

  it('should render correctly when has curve engraving warning', () => {
    mockUseWorkarea.mockReturnValue('fbb2');
    mockUseHasCurveEngraving.mockReturnValue(true);

    const { container } = render(
      <ConfigPanelContext.Provider
        value={{
          dispatch: mockDispatch,
          initState: mockInitState,
          selectedLayers: mockSelectedLayers,
          state: mockContextState as any,
        }}
      >
        <SpeedBlock />
      </ConfigPanelContext.Provider>,
    );

    expect(container).toMatchSnapshot();
    expect(container.querySelector('.warning')).toBeInTheDocument();
  });

  it('should render correctly when has auto feeder vector speed warning', () => {
    mockUseWorkarea.mockReturnValue('fbb2');
    mockUseAutoFeeder.mockReturnValue(true);

    const { container } = render(
      <LayerPanelContext.Provider value={{ hasVector: true } as any}>
        <ConfigPanelContext.Provider
          value={{
            dispatch: mockDispatch,
            initState: mockInitState,
            selectedLayers: mockSelectedLayers,
            state: mockContextState as any,
          }}
        >
          <SpeedBlock />
        </ConfigPanelContext.Provider>
      </LayerPanelContext.Provider>,
    );

    expect(container).toMatchSnapshot();
    expect(container.querySelector('.warning')).toBeInTheDocument();
  });

  it('should render correctly when has low speed warning', () => {
    mockUseWorkarea.mockReturnValue('fhexa1');

    const state = { ...mockContextState, speed: { value: 1 } };
    const { container } = render(
      <LayerPanelContext.Provider value={{ hasVector: true } as any}>
        <ConfigPanelContext.Provider
          value={{
            dispatch: mockDispatch,
            initState: mockInitState,
            selectedLayers: mockSelectedLayers,
            state: state as any,
          }}
        >
          <SpeedBlock />
        </ConfigPanelContext.Provider>
      </LayerPanelContext.Provider>,
    );

    expect(container).toMatchSnapshot();
    expect(container.querySelector('.warning')).toBeInTheDocument();
  });

  test('onChange should work', () => {
    const { container } = render(
      <ConfigPanelContext.Provider
        value={{
          dispatch: mockDispatch,
          initState: mockInitState,
          selectedLayers: mockSelectedLayers,
          state: mockContextState as any,
        }}
      >
        <SpeedBlock />
      </ConfigPanelContext.Provider>,
    );

    expect(mockCreateEventEmitter).toBeCalledTimes(1);
    expect(mockCreateEventEmitter).toHaveBeenLastCalledWith('time-estimation-button');

    expect(mockDispatch).not.toBeCalled();
    expect(mockWriteData).not.toBeCalled();
    expect(mockEmit).not.toBeCalled();
    expect(mockBatchCommand).not.toBeCalled();
    expect(batchCmd.count).toBe(0);

    const input = container.querySelector('input');

    fireEvent.change(input, { target: { value: '88' } });
    expect(mockDispatch).toBeCalledTimes(1);
    expect(mockDispatch).toHaveBeenLastCalledWith({
      payload: { configName: 'CUSTOM_PRESET_CONSTANT', speed: 88 },
      type: 'change',
    });
    expect(mockBatchCommand).toBeCalledTimes(1);
    expect(mockBatchCommand).lastCalledWith('Change speed');
    expect(batchCmd.count).toBe(1);
    expect(mockWriteData).toBeCalledTimes(4);
    expect(mockWriteData).toHaveBeenNthCalledWith(1, 'layer1', 'speed', 88, {
      applyPrinting: true,
      batchCmd,
    });
    expect(mockWriteData).toHaveBeenNthCalledWith(2, 'layer1', 'configName', 'CUSTOM_PRESET_CONSTANT', { batchCmd });
    expect(mockWriteData).toHaveBeenNthCalledWith(3, 'layer2', 'speed', 88, {
      applyPrinting: true,
      batchCmd,
    });
    expect(mockWriteData).toHaveBeenNthCalledWith(4, 'layer2', 'configName', 'CUSTOM_PRESET_CONSTANT', { batchCmd });
    expect(mockEmit).toBeCalledTimes(1);
    expect(mockEmit).toHaveBeenLastCalledWith('SET_ESTIMATED_TIME', null);
    expect(batchCmd.onAfter).toBe(mockInitState);
    expect(mockAddCommandToHistory).toBeCalledTimes(1);
    expect(mockAddCommandToHistory).lastCalledWith(batchCmd);
  });

  test('onChange of value display should work correctly', () => {
    const { getByText } = render(
      <ConfigPanelContext.Provider
        value={{
          dispatch: mockDispatch,
          initState: mockInitState,
          selectedLayers: mockSelectedLayers,
          state: mockContextState as any,
        }}
      >
        <SpeedBlock type="modal" />
      </ConfigPanelContext.Provider>,
    );

    expect(mockCreateEventEmitter).toBeCalledTimes(1);
    expect(mockCreateEventEmitter).toHaveBeenLastCalledWith('time-estimation-button');
    expect(getByText('type: modal')).toBeInTheDocument();

    expect(mockDispatch).not.toBeCalled();
    expect(mockWriteData).not.toBeCalled();
    expect(mockEmit).not.toBeCalled();
    fireEvent.click(getByText('MockConfigValueDisplayButton'));
    expect(mockDispatch).toBeCalledTimes(1);
    expect(mockDispatch).toHaveBeenLastCalledWith({
      payload: { configName: 'CUSTOM_PRESET_CONSTANT', speed: 88 },
      type: 'change',
    });
    expect(mockWriteData).not.toBeCalled();
    expect(mockBatchCommand).not.toBeCalled();
  });
});
