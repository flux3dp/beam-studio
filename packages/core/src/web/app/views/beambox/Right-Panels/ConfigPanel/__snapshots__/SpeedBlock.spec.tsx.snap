// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test SpeedBlock should render correctly when has curve engraving warning 1`] = `
<div>
  <div
    class="panel default"
  >
    <span
      class="title"
    >
      Speed
    </span>
    <div>
      MockConfigValueDisplay
      <p>
        inputId: 
        speed-input
      </p>
      <p>
        type: 
        default
      </p>
      <p>
        max: 
        900
      </p>
      <p>
        min: 
        0.5
      </p>
      <p>
        value: 
        87
      </p>
      <p>
        unit: 
        mm/s
      </p>
      <p>
        hasMultiValue: 
      </p>
      <p>
        decimal: 
        1
      </p>
      <p>
        options: 
        null
      </p>
      <button
        type="button"
      >
        MockConfigValueDisplayButton
      </button>
    </div>
    <input
      class="mock-config-slider"
      id="speed"
      max="900"
      min="0.5"
      step="0.1"
      type="range"
      value="87"
    />
    <div
      class="warning"
    >
      <div
        class="warning-icon"
      >
        !
      </div>
      <div
        class="warning-text"
      >
        The engraving speed of 3D curve objects will be constrained to 50 mm/s. You can remove this limit at Preferences Settings.
      </div>
    </div>
  </div>
</div>
`;

exports[`test SpeedBlock should render correctly when has low speed warning 1`] = `
<div>
  <div
    class="panel default"
  >
    <span
      class="title"
    >
      Speed
    </span>
    <div>
      MockConfigValueDisplay
      <p>
        inputId: 
        speed-input
      </p>
      <p>
        type: 
        default
      </p>
      <p>
        max: 
        900
      </p>
      <p>
        min: 
        0.5
      </p>
      <p>
        value: 
        1
      </p>
      <p>
        unit: 
        mm/s
      </p>
      <p>
        hasMultiValue: 
      </p>
      <p>
        decimal: 
        1
      </p>
      <p>
        options: 
        null
      </p>
      <button
        type="button"
      >
        MockConfigValueDisplayButton
      </button>
    </div>
    <input
      class="mock-config-slider"
      id="speed"
      max="900"
      min="0.5"
      step="0.1"
      type="range"
      value="1"
    />
    <div
      class="warning"
    >
      <div
        class="warning-icon"
      >
        !
      </div>
      <div
        class="warning-text"
      >
        Low speed might cause burning of the material.
      </div>
    </div>
  </div>
</div>
`;

exports[`test SpeedBlock should render correctly when has vector warning 1`] = `
<div>
  <div
    class="panel default"
  >
    <span
      class="title"
    >
      Speed
    </span>
    <div>
      MockConfigValueDisplay
      <p>
        inputId: 
        speed-input
      </p>
      <p>
        type: 
        default
      </p>
      <p>
        max: 
        300
      </p>
      <p>
        min: 
        0.5
      </p>
      <p>
        value: 
        87
      </p>
      <p>
        unit: 
        mm/s
      </p>
      <p>
        hasMultiValue: 
      </p>
      <p>
        decimal: 
        1
      </p>
      <p>
        options: 
        null
      </p>
      <button
        type="button"
      >
        MockConfigValueDisplayButton
      </button>
    </div>
    <input
      class="mock-config-slider"
      id="speed"
      max="300"
      min="0.5"
      step="0.1"
      type="range"
      value="87"
    />
    <div
      class="warning"
    >
      <div
        class="warning-icon"
      >
        !
      </div>
      <div
        class="warning-text"
      >
        The cutting speed of vector path objects will be constrained to 20 mm/s. You can remove this limit at Preferences Settings.
      </div>
    </div>
  </div>
</div>
`;

exports[`test SpeedBlock should render correctly when type is panel-item 1`] = `
<div>
  <div>
    MockObjectPanelItem
    <div>
      id: 
      speed
    </div>
    <div>
      content: 
      <button
        class="adm-button adm-button-default adm-button-fill-outline adm-button-mini adm-button-shape-rounded number-item display-btn"
        type="button"
      >
        <span>
          <span
            style="white-space: nowrap;"
          >
            87
          </span>
        </span>
      </button>
    </div>
    <div>
      label: 
      Speed
    </div>
  </div>
</div>
`;

exports[`test SpeedBlock should render correctly when unit is inches 1`] = `
<div>
  <div
    class="panel default"
  >
    <span
      class="title"
    >
      Speed
    </span>
    <div>
      MockConfigValueDisplay
      <p>
        inputId: 
        speed-input
      </p>
      <p>
        type: 
        default
      </p>
      <p>
        max: 
        300
      </p>
      <p>
        min: 
        0.5
      </p>
      <p>
        value: 
        87
      </p>
      <p>
        unit: 
        in/s
      </p>
      <p>
        hasMultiValue: 
      </p>
      <p>
        decimal: 
        2
      </p>
      <p>
        options: 
        null
      </p>
      <button
        type="button"
      >
        MockConfigValueDisplayButton
      </button>
    </div>
    <input
      class="mock-config-slider"
      id="speed"
      max="300"
      min="0.5"
      step="0.1"
      type="range"
      value="87"
    />
  </div>
</div>
`;

exports[`test SpeedBlock should render correctly when unit is mm 1`] = `
<div>
  <div
    class="panel default"
  >
    <span
      class="title"
    >
      Speed
    </span>
    <div>
      MockConfigValueDisplay
      <p>
        inputId: 
        speed-input
      </p>
      <p>
        type: 
        default
      </p>
      <p>
        max: 
        300
      </p>
      <p>
        min: 
        0.5
      </p>
      <p>
        value: 
        87
      </p>
      <p>
        unit: 
        mm/s
      </p>
      <p>
        hasMultiValue: 
      </p>
      <p>
        decimal: 
        1
      </p>
      <p>
        options: 
        null
      </p>
      <button
        type="button"
      >
        MockConfigValueDisplayButton
      </button>
    </div>
    <input
      class="mock-config-slider"
      id="speed"
      max="300"
      min="0.5"
      step="0.1"
      type="range"
      value="87"
    />
  </div>
</div>
`;
